File: medbrains/apps/ios/MedBrainsStaff/Features/Reception/ReceptionApi.swift
import Foundation
import MedBrainsKit

/// A patient as the desk needs them: the two identifiers (name and date of
/// birth), the phone, and the safety flags. The list, the match and the
/// created record all decode into this.
struct PatientSummary: Decodable, Hashable, Identifiable, Sendable {
    let id: String
    let uhid: String
    let firstName: String
    let lastName: String
    let dateOfBirth: String?
    let phone: String
    let gender: String
    var isMedicoLegal: Bool? = nil
    var mlcNumber: String? = nil
    var isVip: Bool? = nil
    var isDobEstimated: Bool? = nil
    var lastVisitDate: String? = nil
    var totalVisits: Int? = nil
    var outstandingBalance: String? = nil

    var fullName: String { "\(firstName) \(lastName)" }
    /// "UHID · DOB · phone": the line the desk reads back.
    var identifiers: String { [uhid, dateOfBirth ?? "DOB unknown", phone].joined(separator: " · ") }
}

/// A possible duplicate from `POST /api/patients/match`, scored by the server.
struct MatchCandidate: Decodable, Identifiable, Sendable {
    let id: String
    let uhid: String
    let firstName: String
    let lastName: String
    let dateOfBirth: String?
    let phone: String
    let score: Double

    var asSummary: PatientSummary {
        PatientSummary(id: id, uhid: uhid, firstName: firstName, lastName: lastName, dateOfBirth: dateOfBirth, phone: phone, gender: "unknown")
    }
}

struct MatchQuery: Encodable {
    let firstName: String
    let lastName: String
    let dateOfBirth: String?
    let phone: String
}

/// What the phone records about how this registration was checked, so a
/// second UHID for the same person is never silent.
struct MobileRegistrationAttributes: Encodable {
    let source: String
    let duplicateCheck: String
    let matchedUhid: String?
}

struct CreatePatientBody: Encodable {
    struct Attributes: Encodable { let mobileRegistration: MobileRegistrationAttributes }
    let firstName: String
    let lastName: String
    let gender: String
    let phone: String
    let dateOfBirth: String
    let isDobEstimated: Bool
    let registrationType: String
    let registrationSource: String
    let referredByName: String?
    let departmentId: String?
    let consultantId: String?
    let abhaNumber: String?
    let isMedicoLegal: Bool
    let mlcNumber: String?
    let isVip: Bool
    let attributes: Attributes
}

struct DepartmentRow: Decodable, Hashable, Identifiable, Sendable {
    let id: String
    let code: String
    let name: String
}

struct DoctorRow: Decodable, Hashable, Identifiable, Sendable {
    let id: String
    let fullName: String
    let specialization: String?
}

struct StartVisitBody: Encodable {
    let patientId: String
    let departmentId: String
    let doctorId: String?
    let chiefComplaint: String?
    let visitType: String
}

/// The visit and the token the server issued for it, in one answer.
struct VisitStarted: Decodable, Hashable, Sendable {
    struct Encounter: Decodable, Hashable, Sendable { let id: String }
    struct Queue: Decodable, Hashable, Sendable { let id: String; let tokenNumber: Int; let status: String }
    let encounter: Encounter
    let queue: Queue
}

private struct PatientPage: Decodable { let patients: [PatientSummary] }

struct ReceptionApi: Sendable {
    let client: ApiClient

    func matchPatients(_ q: MatchQuery) async throws -> [MatchCandidate] { try await client.request(.post, "/api/patients/match", body: q) }
    func createPatient(_ body: CreatePatientBody) async throws -> PatientSummary { try await client.request(.post, "/api/patients", body: body) }
    func searchPatients(_ text: String) async throws -> [PatientSummary] {
        let q = text.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? text
        let page: PatientPage = try await client.request(.get, "/api/patients?search=\(q)&per_page=25")
        return page.patients
    }
    func departments() async throws -> [DepartmentRow] { try await client.request(.get, "/api/setup/departments") }
    func doctors() async throws -> [DoctorRow] { try await client.request(.get, "/api/setup/doctors") }
    func startVisit(_ body: StartVisitBody) async throws -> VisitStarted { try await client.request(.post, "/api/opd/encounters", body: body) }
}

File: medbrains/apps/android/app-staff/src/main/kotlin/com/medbrains/staff/reception/ReceptionApi.kt
package com.medbrains.staff.reception

import com.medbrains.kit.ApiClient
import kotlinx.serialization.Serializable
import java.net.URLEncoder

/** A patient as the desk needs them: the two identifiers, the phone, the safety flags. */
@Serializable
data class PatientSummary(
    val id: String,
    val uhid: String,
    val first_name: String,
    val last_name: String,
    val date_of_birth: String? = null,
    val phone: String,
    val gender: String,
    val is_medico_legal: Boolean? = null,
    val mlc_number: String? = null,
    val is_vip: Boolean? = null,
    val is_dob_estimated: Boolean? = null,
    val last_visit_date: String? = null,
    val total_visits: Int? = null,
    val outstanding_balance: String? = null,
) {
    val fullName: String get() = "$first_name $last_name"
    /** "UHID · DOB · phone": the line the desk reads back. */
    val identifiers: String get() = listOf(uhid, date_of_birth ?: "DOB unknown", phone).joinToString(" · ")
}

/** A possible duplicate from `POST /api/patients/match`, scored by the server. */
@Serializable
data class MatchCandidate(val id: String, val uhid: String, val first_name: String, val last_name: String, val date_of_birth: String? = null, val phone: String, val score: Double) {
    fun asSummary() = PatientSummary(id, uhid, first_name, last_name, date_of_birth, phone, "unknown")
}

@Serializable data class MatchQuery(val first_name: String, val last_name: String, val date_of_birth: String?, val phone: String)

/** What the phone records about how this registration was checked, so a second UHID is never silent. */
@Serializable data class MobileRegistration(val source: String, val duplicate_check: String, val matched_uhid: String?)
@Serializable data class RegistrationAttributes(val mobile_registration: MobileRegistration)

@Serializable
data class CreatePatientBody(
    val first_name: String,
    val last_name: String,
    val gender: String,
    val phone: String,
    val date_of_birth: String,
    val is_dob_estimated: Boolean,
    val registration_type: String,
    val registration_source: String,
    val referred_by_name: String?,
    val department_id: String?,
    val consultant_id: String?,
    val abha_number: String?,
    val is_medico_legal: Boolean,
    val mlc_number: String?,
    val is_vip: Boolean,
    val attributes: RegistrationAttributes,
)

@Serializable data class DepartmentRow(val id: String, val code: String, val name: String)
@Serializable data class DoctorRow(val id: String, val full_name: String, val specialization: String? = null)
@Serializable data class StartVisitBody(val patient_id: String, val department_id: String, val doctor_id: String?, val chief_complaint: String?, val visit_type: String)

/** The visit and the token the server issued for it, in one answer. */
@Serializable
data class VisitStarted(val encounter: Encounter, val queue: Queue) {
    @Serializable data class Encounter(val id: String)
    @Serializable data class Queue(val id: String, val token_number: Int, val status: String)
}

@Serializable private data class PatientPage(val patients: List<PatientSummary>)

class ReceptionApi(private val client: ApiClient) {
    suspend fun matchPatients(q: MatchQuery): List<MatchCandidate> = client.post("/api/patients/match", q)
    suspend fun createPatient(body: CreatePatientBody): PatientSummary = client.post("/api/patients", body)
    suspend fun searchPatients(text: String): List<PatientSummary> =
        client.get<PatientPage>("/api/patients?search=${URLEncoder.encode(text, "UTF-8")}&per_page=25").patients
    suspend fun departments(): List<DepartmentRow> = client.get("/api/setup/departments")
    suspend fun doctors(): List<DoctorRow> = client.get("/api/setup/doctors")
    suspend fun startVisit(body: StartVisitBody): VisitStarted = client.post("/api/opd/encounters", body)
}
